accepted/000071-cluster_awareness.md

```diff
@@ -0,0 +1,322 @@
+
+Future developments can enrich the implementations described in the RFC by converting the cluster management node into a [Salt proxy minion](https://docs.saltstack.com/en/latest/topics/proxyminion/index.html) using the SSH backend to access all cluster nodes.
+In this way, the cluster object can be represented by a special minion registered to Uyuni/SUSE Manager and the user can issue some operations on the cluster object that will translate the action to all cluster nodes and collect the results. Some examples:
+
+* Assign Salt states
+* Execute Salt execution modules
+* Assign software channels and manage Channel Lifecycle Management
+
+We listed this feature as a future development because:
+* The MVP actions we outlined above have higher priority
+* During our experiments we encountered [this bug](https://github.com/saltstack/salt/issues/53341) with `salt-proxy`, we correctly got past it but we were still unable to register a proxied system to the `salt-master`
+* Some of the features can be already implemented by scripting and using Salt on the cluster.
+
+## SSH keys management
+
+Any SSH key pair required for accessing the cluster nodes will be stored on the server.
+
+The keys will be made available to the provisioning node by:
+1. loading them into the `ssh-agent` on the server
+2. opening a connection to the provisioning node using agent forwarding (`ssh -A`)
+
+The connection must be opened before executing any action that might require connecting to the cluster nodes. E.g. joining a node, getting the cluster status, etc.
+The server must retrieve the value of the `SSH_AUTH_SOCK` environment variable and inject it into subsequent `state.apply`s as a pillar.
+
+The salt state corresponding to the action must ensure the `SSH_AUTH_SOCK` environment variable is set correctly to allow access to the forwarded agent.
+
+Once the action is completed the SSH connection can be terminated.
+
+E.g. a state that calls `skuba` to get the status of a CaaSP cluster:
+
+```yaml
+ssh_agent_socket:
+    environ.setenv:
+        - name: SSH_AUTH_SOCK
+        - value: {{ pillar['ssh_auth_sock'] }}
+
+cluster_status:
+     cmd.run:
+         - name: skuba cluster status
+         - cwd: {{ pillar['cluster_dir'] }}
+         - require:
+           - environ: ssh_agent_socket
+...
+```
+
+```
+salt <provisioning.node> state.apply caasp.join pillar='{"ssh_auth_sock": "/tmp/....", ...}'
+```
+
+## Pluggable architecture
+
+The MVP will not be delivered as a plugin for the following reasons:
+- longer development time for implementing an MVP
+- unforeseen technical problems related to plugins. We don't expect anything major but small problems can add up to the overall development time.
+- it will result in a large and complex pull request which will require a lot of time and effort for reviewing
+
+After the first product is integrated into Uyuni/SUSE Manager, it can be extracted into a plugin and make the whole cluster architecture pluggable.
```

Comment:
Agreed on this one. BTW, the more I'm thinking about it, the more obvious it becomes that there will be a ton of overlaps between how we're handling existing "external components" like monitoring servers, VM hosts, and build hosts and how those cluster management admin servers will work. So I'd expect us to be able to re-use the lessons learned to actually make those more traditional "add-ons" just as pluggable.

Reply:
+1